src/Storages/MergeTree/Compaction/MergeSelectors/SimpleMergeSelector.cpp

```diff
@@ -312,12 +318,12 @@ PartsRanges SimpleMergeSelector::select(
 
     /// Using max size constraint to create more merge candidates
     for (auto range_it = parts_ranges.begin(); range_it != parts_ranges.end(); ++range_it)
-        selectWithinPartsRange(range_it, max_merge_sizes[0], range_filter, estimator, settings, min_size_to_lower_base_log, max_size_to_lower_base_log);
+        selectWithinPartsRange(range_it, merge_constraints[0], range_filter, estimator, settings, min_size_to_lower_base_log, max_size_to_lower_base_log);
 
     PartsRanges result;
-    for (size_t max_merge_size : max_merge_sizes)
+    for (const auto & constraint : merge_constraints)
     {
-        if (auto range = estimator.buildMergeRange(max_merge_size))
+        if (auto range = estimator.buildMergeRange(constraint.max_size_bytes))
```

Comment:
You need to include row constraints into `buildMergeRange`, otherwise estimator will return partially filtered results.